docs/proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,189 @@
+Custom Metrics API
+==================
+
+The new [metrics monitoring vision]() proposes an API that the Horizontal
+Pod Autoscaler can use to access arbitrary metrics.
+
+Similarly to the [master metrics API](resource-metrics-api.md), the new
+API should be structured around accessing metrics by referring to
+kubernetes objects (or groups thereof) and a metric name.  For this
+reason, the API could be useful for other consumers (most likely
+controllers) that want to consume custom metrics (similarly to how the
+master metrics API is generally useful to multiple cluster components).
+
+The HPA can currently refer to all pods selected by a label selector, and
+is planned to be able to support referring to metrics describing
+a namespace and other non-pod objects within the namespace.
+
+API Paths
+---------
+
+The root API path will look like `/apis/general-metrics/v1alpha1`.  For
+brevity, this will be left off below.
+
+- `/nodes/{node-name}/metrics/{metric-name...}`: retrieve the given metric
+  on the given node
+
+- `/namespaces/{namespace-name}/metrics/{metric-name...}`: retrieve the
+  given metric on the given namespace
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}`:
+  retrieve the given metric for all objects of the given type in the given
+  namespace.
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}?labelSelector=foo`:
+  retrieve the given metric for all objects of the given type matching the
+  given label selector in the given namespace.
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}?names=foo,bar`:
+  retrieve the given metric for the objects of the given type with the
+  given name in the given namespace
+
+For example, to retrieve the custom metric "hits-per-second" for all pods
+matching "app=frontend` in the namespaces "webapp", the request might look
+like:
+
+`/apis/general-metrics/v1alpha1/namespaces/webapp/object-metrics/pods/hits-per-second?labelSelector=app%3Dfrontend`.
+
+API Objects
```

Comment:
> Is this API supposed to return the most recent value of a given metric?

Yes, just like the master metrics API.

> In HPA we want to scale if number of all 200s is above 57 per instance. How can I express this using this API?

> How does it combine multiple timeseries?

The HPA v2 has no notion of labels, so I did not put any notion of labels into this API.  My initial gut reaction is to say that the adapter should have a way to expose labeled metrics within this schema, but I'm open to ideas.